### ISSUE TITLE
Shader language: Implement Struct Support

### ISSUE THREAD

**Issue by TheRealNOIG** [godotengine/godot-proposals#301]
**Describe the project you are working on:** I'm currently working on porting all the shaders from my OpenGL engine to Godot and have notices a major missing feature that doesn't have a request made for it already that is used extensively in animated shaders

**Describe the problem or limitation you are having in your project:** I am unable to make data structures in the current implementation of Godots shaders.

**Describe how this feature / enhancement will help you overcome this problem or limitation:** Structs is one of GLSL's most powerful feature allowing for a large amount of functionality(Its not just used for animation but this will be the example I give).  Structs are needed for creating and tracking animated objects within the fragment shader. This allows for an absolutely large amount of animated textures to be created. Here is just one example I found on shadertoy.com Credit: AntoineC https://www.shadertoy.com/view/MsGczV. If you browse that site you will find that structs are used extensively. That is just one of the uses for structs. (There is endless usefulness having structured data that can be dynamically created and manipulated in any programing language)

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:** 
GLSL 
```
struct Light
{
  vec3 eyePosOrDir;
  bool isDirectional;
  vec3 intensity;
  float attenuation;
} variableName;
```
And the animated texture showed [above](https://www.shadertoy.com/view/MsGczV)

**Describe implementation detail for your proposal (in code), if possible:**

**If this enhancement will not be used often, can it be worked around with a few lines of script?:** 
No. There might be some work around in the future once arrays are added to Godots' shader program [#10751](https://github.com/godotengine/godot/issues/10751)

**Is there a reason why this should be core and not an add-on in the asset library?:** I dont know how the add-on system works in Godot but I wouldn't guess that this would work within that system.

**Comment by Hidden-Abstraction** [comment]
I hope that for v4.0 the missing features in the shader language are added.

I have a lot of code written in hlsl (Unity) that I still can't include in godot because of the missing features.

**Comment by SIsilicon** [comment]
I don't know why, but this doesn't seem to be working for me. I'm using Godot 3.2.1 on GLES3.
![image](https://user-images.githubusercontent.com/34734122/80515585-ca6c4d00-8947-11ea-9bba-22e3d0a3d011.png)

**Comment by clayjohn** [comment]
@SIsilicon Yes, it was implemented for 4.0 only.

**Comment by EvgeneKuklin** [comment]
When I try to do like this:

```glsl
shader_type spatial;

struct Layer {
    sampler2D foreground_texture;
}
```
I got an error:
`A 'sampler2D' data type is not allowed here.`

Why?
Is there a way around it?
